### PR TITLE
#556 Strict indentation check is wrong

### PR DESCRIPTION
**Pull request by HDouss**
Resolving #556 Strict indentation check is wrong.
Added 2 unit tests.
Exception to the rule described in #556 : Lines in block comments

### CODE REVIEW

**Comment by HDouss** [comment]
@davvd PR ready. Please assign a reviewer.

**Comment by davvd** [comment]
@hdouss Thanks, I will find someone to review this PR soon

**Comment by davvd** [comment]
@pinaf review this one, please

File: qulice-checkstyle/src/main/java/com/qulice/checkstyle/QuliceIndentationCheck.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    /**
+     * {@inheritDoc}
+     * @checkstyle MagicNumberCheck (13 lines)
+     */
+    @Override
+    public void processFiltered(final File file, final List<String> lines) {
+        int previous = 0;
+        for (int pos = 0; pos < lines.size(); pos += 1) {
+            final String line = lines.get(pos);
+            final int current = indentation(line);
```
Comment (pinaf) [inline qulice-checkstyle/src/main/java/com/qulice/checkstyle/QuliceIndentationCheck.java]: @HDouss `QuliceIndentationCheck.indentation(line);`

File: qulice-checkstyle/src/main/java/com/qulice/checkstyle/QuliceIndentationCheck.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+            if (inCommentBlock(line) || line.length() == 0) {
```
Comment (pinaf) [inline qulice-checkstyle/src/main/java/com/qulice/checkstyle/QuliceIndentationCheck.java]: @HDouss `QuliceIndentationCheck.inCommentBlock`

File: qulice-checkstyle/src/main/java/com/qulice/checkstyle/QuliceIndentationCheck.java [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+     * @param line Input.
+     * @return True if the line belongs to a comment block.
+     */
+    private static boolean inCommentBlock(final String line) {
+        final String trimmed = line.trim();
+        return trimmed.length() > 0
+            && (trimmed.charAt(0) == '*'
+                || trimmed.startsWith("/*")
+                || trimmed.startsWith("*/")
+                );
```
Comment (pinaf) [inline qulice-checkstyle/src/main/java/com/qulice/checkstyle/QuliceIndentationCheck.java]: @HDouss what about comments that start with `//`?

**Comment by pinaf** [comment]
@HDouss 3 comments

File: qulice-checkstyle/src/main/java/com/qulice/checkstyle/QuliceIndentationCheck.java (hunk 3 above)
Comment (HDouss) [inline qulice-checkstyle/src/main/java/com/qulice/checkstyle/QuliceIndentationCheck.java]: @pinaf Unlike lines in comment block, single line comments should follow the rule. The rule is even forced by Checkstyle.

**Comment by HDouss** [comment]
@pinaf Thanks. I just added a commit for your two first comments.

**Comment by pinaf** [comment]
@rultor merge

**Comment by pinaf** [comment]
@HDouss thanks!

**Comment by HDouss** [comment]
@krzyk Please ask for merge

File: qulice-checkstyle/src/main/java/com/qulice/checkstyle/QuliceIndentationCheck.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+ * <ul>
+ * <li>the same as previous one or less
+ * <li>bigger than previous by exactly 4
+ * </ul>
+ * All other cases must cause a failure.
+ *
+ * @author Hamdi Douss (douss.hamdi@gmail.com)
+ * @version $Id$
+ */
+public final class QuliceIndentationCheck extends AbstractFileSetCheck {
```
Comment (krzyk) [inline qulice-checkstyle/src/main/java/com/qulice/checkstyle/QuliceIndentationCheck.java]: @HDouss please rename it to `CascadeIndentationCheck`

File: qulice-checkstyle/src/main/java/com/qulice/checkstyle/QuliceIndentationCheck.java [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        int previous = 0;
+        for (int pos = 0; pos < lines.size(); pos += 1) {
+            final String line = lines.get(pos);
+            final int current = QuliceIndentationCheck.indentation(line);
+            if (QuliceIndentationCheck.inCommentBlock(line)
+                || line.length() == 0) {
+                continue;
+            }
+            if (current > previous
+                && current != previous + 4) {
```
Comment (krzyk) [inline qulice-checkstyle/src/main/java/com/qulice/checkstyle/QuliceIndentationCheck.java]: @HDouss put `4` in a constant and then you can reuse it in the string below, and remove the `MagicNumberCheck` exclude

**Comment by krzyk** [comment]
@HDouss please see my comments

**Comment by HDouss** [comment]
@krzyk Thanks ! Done.

**Comment by krzyk** [comment]
@HDouss one more and we are good

**Comment by HDouss** [comment]
@krzyk Won't qulice complain without it ?

**Comment by HDouss** [comment]
@krzyk I just checked. Checkstyle complain with : 
`CascadeIndentationCheck.java[50]: Missing a Javadoc comment.`

**Comment by krzyk** [comment]
@HDouss no, `@Overriden` methods don't need it

**Comment by krzyk** [comment]
@HDouss strange

**Comment by HDouss** [comment]
@krzyk Sorry. I thought you are talking about javadoc comment for the constant.

**Comment by HDouss** [comment]
@krzyk Thanks for your review. Done !

**Comment by krzyk** [comment]
@HDouss thanks, I checked updated qulice on one of the projects and it works great :)

**Comment by krzyk** [comment]
@rultor merge pls

**Comment by davvd** [comment]
@pinaf I added **21 mins** to your account, many thanks for working with the project (AP-0PB81359M7917031K)! The completion time here was 28 hours and 33 mins.; you're getting extra minutes here (c=6); added +21 to your rating, now it is equal to [+4990](http://www.netbout.com/b/35128?open=rating)

**Comment by davvd** [comment]
@rultor deploy now pls

**Pull request merged**